sdk/core/System.ClientModel/src/Internal/ReadTimeoutStream.cs

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Primitives;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Internal;
+
+/// <summary>
+/// Read-only Stream that will throw a <see cref="OperationCanceledException"/> if it has to wait longer than a configurable timeout to read more data
+/// </summary>
+internal class ReadTimeoutStream : Stream
+{
+    private readonly Stream _stream;
+    private TimeSpan _readTimeout;
+    private CancellationTokenSource _cancellationTokenSource = null!;
+
+    public ReadTimeoutStream(Stream stream, TimeSpan readTimeout)
+    {
+        _stream = stream;
+        _readTimeout = readTimeout;
+
+        UpdateReadTimeout();
+        InitializeTokenSource();
+    }
+
+    public override bool CanRead => _stream.CanRead;
+
+    public override bool CanSeek => _stream.CanSeek;
+
+    public override bool CanWrite => false;
+
+    public override long Length => _stream.Length;
+
+    public override long Position
+    {
+        get => _stream.Position;
+        set => _stream.Position = value;
+    }
+
+    public override int ReadTimeout
+    {
+        get => (int)_readTimeout.TotalMilliseconds;
+        set
+        {
+            _readTimeout = TimeSpan.FromMilliseconds(value);
+            UpdateReadTimeout();
+        }
+    }
+
+    public override void Close()
+        => _stream.Close();
+
+    public override void Flush()
+    {
+        // Flush is allowed on read-only stream
+    }
+
+    public override int Read(byte[] buffer, int offset, int count)
+    {
+        var source = StartTimeout(default, out bool dispose);
+        try
+        {
+            return _stream.Read(buffer, offset, count);
```

Comment:
Is it intentional that we don't throw until after Read returns, even if it takes longer than the timeout?

Reply:
I think the difficulty is that we don't know the `count` value that this method would be called with. In ResponseBodyPolicy we implement timeouts within the sync path when reading from a stream, but in that case we control the chunk size and we use 81920. So assuming the client library/user code doesn't read in larger chunks than that we would have the same behavior as [ResponseBodyPolicy](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Pipeline/Internal/ResponseBodyPolicy.cs#L146-L149). But if we want to be really safe, we can probably take the min of (81920, count) and if count is greater than 81920, we could read in chunks of 81920 and write to the buffer in chunks.

Reply:
Tracking the request for helpful comments with https://github.com/Azure/azure-sdk-for-net/issues/41305